ui/ozone/platform/starboard/ozone_platform_starboard.cc

```diff
@@ -0,0 +1,154 @@
+    return cursor_factory_.get();
+  }
+
+  ui::InputController* GetInputController() override {
+    return input_controller_.get();
+  }
+
+  ui::GpuPlatformSupportHost* GetGpuPlatformSupportHost() override {
+    return gpu_platform_support_host_.get();
+  }
+
+  std::unique_ptr<SystemInputInjector> CreateSystemInputInjector() override {
+    // TODO(b/371272304): Implement a system input injector for Starboard. This
+    // will likely be part of an event factory/dispatcher.
+    return nullptr;
+  }
+
+  std::unique_ptr<PlatformWindow> CreatePlatformWindow(
+      PlatformWindowDelegate* delegate,
+      PlatformWindowInitProperties properties) override {
+    return std::make_unique<PlatformWindowStarboard>(delegate,
+                                                     properties.bounds);
+  }
+
+  std::unique_ptr<display::NativeDisplayDelegate> CreateNativeDisplayDelegate()
+      override {
+    // TODO(b/371272304): Implement a native display delegate to allow display
+    // configuration and hotplug (adding a new display while the process is
+    // running).
+    return std::make_unique<display::FakeDisplayDelegate>();
+  }
+
+  std::unique_ptr<PlatformScreen> CreateScreen() override {
+    return std::make_unique<PlatformScreenStarboard>();
+  }
+  // This function must be called immediately after CreateScreen with the
+  // `screen` that was returned from CreateScreen. They are separated to avoid
+  // observer recursion into display::Screen from inside CreateScreen.
+  void InitScreen(PlatformScreen* screen) override {}
+
+  std::unique_ptr<InputMethod> CreateInputMethod(
+      ImeKeyEventDispatcher* ime_key_event_dispatcher,
+      gfx::AcceleratedWidget widget) override {
+    return std::make_unique<InputMethodMinimal>(ime_key_event_dispatcher);
+  }
+
+  bool InitializeUI(const InitParams& params) override {
+    // TODO(b/371272304): Investigate if we need a more robust cursor factory or
+    // if we can continue using BitmapCursorFactory.
+    cursor_factory_ = std::make_unique<BitmapCursorFactory>();
+    // TODO(b/371272304): GpuPlatformSupportHost seems to be mainly for IPC
+    // between the Browser and GPU processes. Since Cobalt will be running as a
+    // single process, investigate if this is needed or if there's any
+    // additional features we might need to implement from there.
+    gpu_platform_support_host_.reset(CreateStubGpuPlatformSupportHost());
+    // TODO(b/371272304): Investigate if we need our own implementation of
+    // InputController for things like gamepads or other atypical input devices.
+    input_controller_ = CreateStubInputController();
+    // TODO(b/371272304): Investigate if we need our own implementation of
+    // OverlayManager.
+    if (!overlay_manager_) {
+      overlay_manager_ = std::make_unique<StubOverlayManager>();
+    }
+    if (!surface_factory_) {
+      surface_factory_ = std::make_unique<SurfaceFactoryStarboard>();
+    }
```

Comment:
Headless and Flatland both allow it to be on either process, with headless doing it similar to this where whatever process gets to it first will initialize it. Flatland has a DCHECK to verify it's either initialized by the gpu process or it's only running in a single process, in which case either initialize call is allowed to create it. Since we plan to generally run in only a single process I could add a similar check for us.

ozone_platform_cast also has paths for it to initialize in either process, but with more targeted checks. They only allow it to initialize in the UI or the getter if the "allow-dummy-software-rendering" switch is passed, otherwise they crash because they've reached a point where it should have been initialized in the GPU process but is still null. https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/platform/cast/ozone_platform_cast.cc;l=59

Reply:
Minor correction: headless and flatland would allow _each_ process to  
create their own surface factory and overlay manager (unless 
--single-process). I think then the code is OK as-is, but I'd  
like to leave a comment in each initialization pair, something  
like 
```
// It doesn't matter who creates |overlay_manager_ | and  
// | surface_factory_ | in single-process mode.
```

so that it's clear that this is intentional and not accidental.

Reply:
Done.